### PR TITLE
adding unified matrix interface

### PR DESCRIPTION
**Pull request by chivee**
1. fix bug: sync server will block unexpected #47 
2. adding an unify matrix interface.

### CODE REVIEW

**Comment by chivee** [comment]
@Microsoft/multiverso  please review and signed off

**Comment by chivee** [comment]
We can replace the matrixtable with this one once commited

File: src/table/matrix.cpp [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+
+#include "multiverso/io/io.h"
+#include "multiverso/multiverso.h"
+#include "multiverso/util/log.h"
+#include "multiverso/util/quantization_util.h"
+#include "multiverso/updater/updater.h"
+
+namespace multiverso {
+
+  template <typename T>
```
Comment (feiga) [inline src/table/matrix.cpp]: extra spaces

File: include/multiverso/table_interface.h [hunk 2]
```diff
@@ -30,25 +31,26 @@ class WorkerTable {
...
-  void Reset(int msg_id, int num_wait);
+  void Reset(integer_t msg_id, integer_t num_wait);
 
-  void Notify(int id);
+  void Notify(integer_t id);
 
   virtual int Partition(const std::vector<Blob>& kv,
-    std::unordered_map<int, std::vector<Blob> >* out) = 0;
+   std::unordered_map<int, std::vector<Blob> >* out,
+                             MsgType partition_type) = 0;
```
Comment (feiga) [inline include/multiverso/table_interface.h]: suggest to make out as the last arguments since it is a output argument

File: src/table/matrix.cpp [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+  void MatrixWorker<T>::Get(T* data, size_t size,
+                         const GetOption* option) {
+    CHECK(size == num_col_ * num_row_);
+    integer_t whole_table = -1;
+    Get(whole_table, data, size, option);
+  }
+
+  template <typename T>
+  void MatrixWorker<T>::Get(integer_t row_id, T* data, size_t size,
+                                           const GetOption* option) {
```
Comment (feiga) [inline src/table/matrix.cpp]: align left with first argument?

File: src/table/matrix.cpp [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+
+    WorkerTable::Get(ids_blob, option);
+    Log::Debug("[Get] worker = %d, #rows_set = %d / %d\n", MV_Rank(), row_ids_size, num_row_);
+    if (is_option_mine) delete option;
+  }
+
+  template <typename T>
+  void MatrixWorker<T>::Add(T* data, size_t size, const AddOption* option) {
+    CHECK(size == num_col_ * num_row_);
+    if (is_sparse_ && true) {
```
Comment (feiga) [inline src/table/matrix.cpp]: why && true?

File: src/table/matrix.cpp [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+      }
+
+      bool is_option_mine = false;
+      if (option == nullptr) {
+        is_option_mine = true;
+        option = new AddOption();
+      }
+
+      WorkerTable::Add(ids_blob, data_blob, option);
+      Log::Debug("[Add] Sparse: worker = %d, #rows_set = %d / %d\n", MV_Rank(), row_ids.size(), num_row_);
```
Comment (feiga) [inline src/table/matrix.cpp]: make a line less than 80 characters

File: src/table/matrix.cpp [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+    WorkerTable::Add(ids_blob, data_blob, option);
+    Log::Debug("[Add] worker = %d, #row = %d\n", MV_Rank(), row_id);
+    if (is_option_mine) delete option;
+  }
+
+  template <typename T>
+  void MatrixWorker<T>::Add(const std::vector<integer_t>& row_ids,
+                                  const std::vector<T*>& data_vec,
+                                                      size_t size,
+                                          const AddOption* option) {
```
Comment (feiga) [inline src/table/matrix.cpp]: align left to first argument?

File: src/table/matrix.cpp [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+  template <typename T>
+  void MatrixWorker<T>::ProcessReplyGet(std::vector<Blob>& reply_data) {
+
+
+    size_t keys_size = reply_data[0].size<integer_t>();
+    integer_t* keys = reinterpret_cast<integer_t*>(reply_data[0].data());
+    T* data = reinterpret_cast<T*>(reply_data[1].data());
+
+    if (is_sparse_)
+    {
```
Comment (feiga) [inline src/table/matrix.cpp]: move { to the end of if statement

File: include/multiverso/table_interface.h [hunk 8]
```diff
@@ -30,25 +31,26 @@ class WorkerTable {
   int GetAsync(Blob keys, const GetOption* option = nullptr);
   int AddAsync(Blob keys, Blob values, const AddOption* option = nullptr);
 
-  void Wait(int id);
+  void Wait(integer_t id);
```
Comment (feiga) [inline include/multiverso/table_interface.h]: only use the integer_t for row index?

File: src/table/matrix.cpp [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+    CHECK(row_ids.size() == data_vec.size());
+    for (auto i = 0; i < num_row_ + 1; ++i) row_index_[i] = nullptr;
+    for (auto i = 0; i < row_ids.size(); ++i){
+      row_index_[row_ids[i]] = data_vec[i];
+    }
+
+    bool is_option_mine = false;
+    if (is_sparse_ && option == nullptr) {
+        // the get option is required by the sparse update logic.
+        is_option_mine = true;
```
Comment (feiga) [inline src/table/matrix.cpp]: only two space

File: src/table/matrix.cpp (hunk 4 above)
Comment (chivee) [inline src/table/matrix.cpp]: leave an option for this feature.

Comment (chivee) [inline src/table/matrix.cpp]: for now, the compiler will ignore that condition

File: src/table/matrix.cpp (hunk 3 above)
Comment (chivee) [inline src/table/matrix.cpp]: i think the right alignment for the argument is easy to read

File: include/multiverso/table_interface.h (hunk 2 above)
Comment (chivee) [inline include/multiverso/table_interface.h]: done

File: src/table/matrix.cpp (hunk 1 above)
Comment (chivee) [inline src/table/matrix.cpp]: done

File: include/multiverso/table_interface.h (hunk 8 above)
Comment (chivee) [inline include/multiverso/table_interface.h]: so we don't handle the msg more than 2^32?

**Comment by chivee** [comment]
@feiga Please signed off if there isn't any concerns

File: include/multiverso/table_interface.h (hunk 8 above)
Comment (feiga) [inline include/multiverso/table_interface.h]: id can be re-used, like the TCP sequence number. Moreover, we may don't want it consistent with the row index.

File: src/table/matrix_table.cpp [hunk 10]
```diff
@@ -69,20 +69,20 @@ void MatrixWorkerTable<T>::Get(integer_t row_id, T* data, size_t size) {
...
   WorkerTable::Get(Blob(&row_id, sizeof(integer_t)));
   Log::Debug("[Get] worker = %d, #row = %d\n", MV_Rank(), row_id);
 }
 
 template <typename T>
 void MatrixWorkerTable<T>::Get(const std::vector<integer_t>& row_ids,
-  const std::vector<T*>& data_vec,
-  size_t size) {
+                                     const std::vector<T*>& data_vec,
+                                                         size_t size) {
```
Comment (feiga) [inline src/table/matrix_table.cpp]: keep the style consistent

File: include/multiverso/table_interface.h [hunk 11]
```diff
@@ -30,25 +31,26 @@ class WorkerTable {
...
 
-  void Reset(int msg_id, int num_wait);
+  void Reset(integer_t msg_id, integer_t num_wait);
```
Comment (feiga) [inline include/multiverso/table_interface.h]: num_wait as int

File: include/multiverso/table_interface.h [hunk 12]
```diff
@@ -30,25 +31,26 @@ class WorkerTable {
...
+   std::unordered_map<int, std::vector<Blob> >* out) = 0;
 
   virtual void ProcessReplyGet(std::vector<Blob>&) = 0;
 
   // add user defined data structure
 private:
   std::string table_name_;
   // assuming there are at most 2^32 tables
-  int table_id_;
+  integer_t table_id_;
```
Comment (feiga) [inline include/multiverso/table_interface.h]: I suggest use int. 2^32 is a really large number for table id. Only use the typedef type with specific situation.

File: src/table/matrix.cpp [hunk 13]
```diff
@@ -0,0 +1,584 @@
...
+    }
+
+    bool is_option_mine = false;
+    if (option == nullptr) {
+      is_option_mine = true;
+      option = new AddOption();
+    }
+
+    WorkerTable::Add(ids_blob, data_blob, option);
+    Log::Debug("[Add] Sparse: worker = %d, #rows_set = %d / %d\n", MV_Rank(), row_ids.size(), num_row_);
```
Comment (feiga) [inline src/table/matrix.cpp]: keep the line less than 80 characters.

**Pull request merged**